### ISSUE TITLE
Text gets scrambled

### ISSUE THREAD

**Issue by lsaranto** [OptiKey/OptiKey#639]
# How OptiKey was being used

Type text with Optikey and then change position with a mouse click and then use Optikey to do a backspace. 

# What the user expected to happen

Remove a character.

# What the user actually experienced

a section of the text from the end is moved to the current position

# Version
 - Operating system and version:  Windows 7
 - OptiKey version:  I updated to 3.0.2 only recently and it has happened once already.

# Steps to reproduce the issue

Of course, I can't trigger it now. Sorry.

**Comment by JuliusSweetland** [comment]
@lsaranto It sounds like you moved the cursor in the output text and then used a backspace (back one or back many). The issue is that Optikey only has an idea of what has been output, which is displayed in the Optikey scratchpad (text in blue at the top). If you move the cursor, or change focus to another window, or something along those lines, then Optikey has no idea that something has changed, and still thinks the cursor is in the old position. If you then perform a backspace Optikey will assume that you are wanting to remove one (or many if you have output a suggested word or captured a multi-key selection) character and will output one or many backspaces to achieve what it thinks you want to do. The likely result is that the wrong number of backspaces are output as Optikey has the wrong view of the state of the other application.

If you move the cursor manually I would suggest you clear the optikey scratchpad before continuing. This wipes Optikey's view of the world and prevents unwanted effects.

**Comment by lsaranto** [comment]
@JuliusSweetland What you described what might be happening is pretty much what I thought, too. What I don't get is why it doesn't happen every time. When I was typing the previous post (and now this post) I tried several times and ways to trigger it, but couldn't.  I just thought to report it and maybe it rings a bell with someone.

At least` it doesn't happen too often. It's fine by me if we call it a feature. :smiley: 

I use a dynamic keyboard and the scratchpad isn't visible. So it's a bit inconvenient to go clear the scratchpad every time I jump back to fix an earlier sentence. I also don't have the back-many key visible.

**Comment by JuliusSweetland** [comment]
@lsaranto Ah I understand the use case now. 

@AdamRoden What do you think about a new dynamic keyboard property to suppress the scratchpad text history. Defaulted to false, if set to true then this changes the behaviour of the KeyboardPublishService so that the Text property is not updated or used when BackOne/BackMany are triggered. When the user leaves the keyboard the original Text property (and behaviour) is reset. Thoughts?

**Comment by lsaranto** [comment]
@JuliusSweetland Oh, and to clarify, I'm using eye-tracking (Tobii 4C) when typing with OptiKey.

**Comment by lsaranto** [comment]
Still experiencing this on regular basis. Really inconvenient as I don't know the actual trigger or any way to counter from happening. So I'm always fearing it to trigger.

I can use a mouse a bit, and I do some typing with the mouse and another On-screen keyboard -- especially typo fixes. So when I jump back and forth the text, if I then press BackOne on Optikey, something bad might happen. Just moments ago Optikey deleted a whole message as a result of this behavior.

**Comment by JuliusSweetland** [comment]
@AdamRoden Hey Adam. Do you have any thoughts on this? My first thought is to suppress the scratchpad (Text property in KeyboardOutputService), but a better approach may be another property in the KeyboardOutputService which enforces that BackOne is always a single character. This could be linked to whether a dynamic keyboard shows a scratchpad or not. We could optionally add a new setting so 'BackOne always removes a single character'?

**Comment by kmcnaught** [comment]
@lsaranto is this issue still current?

**Comment by lsaranto** [comment]
@kmcnaught Yes, it still hits me maybe once a week, even though I try to remember not to use OptiKey backspace when editing typos or changing letters in the middle of a sentence. Still haven't found a sure way to reproduce the issue. 

Note: In reference to another recent (now closed) issue I posted, I reverted back to version 3.1.0 so I have minimal experience with 3.2.x in regards to this issue.

**Comment by lsaranto** [comment]
Just to give ideas on how to counter this issue: The other on-screen keyboard that I use with a mouse has a word suggestion feature. When typing letters it shows a list of word suggestions. This buffer is cleared if I press a non-character key (eg. END, PGDN, etc.) or click with the mouse cursor outside the on-screen keyboard window. That's a pretty clear indication that that particular word can be forgotten. When I start typing again, it treats it as a new word.

Of course, OptiKeys buffer (scratchpad) covers more than a single word so maybe this won't work so easily.

**Comment by kmcnaught** [comment]
That's a good idea, we can have a think about whether that could be applied to the scratchpad without it giving unexpected behaviour in the context of multiple words. Perhaps at a minimum those non-character keys/mouse events could switch the scratchpad back to an enforced "backone is only allowed to remove a single character" mode, but you might still get strange interactions with word suggestions. I'll have a think.

**Comment by kmcnaught** [comment]
@lsaranto There is one other scenario where `BackOne` does more than remove a single character, and that is when the current word contains some diacritics or other characters that are not in normalized Unicode form. In your first report you mention that `a section of the text from the end is moved to the current position` which sounds exactly like what I'd expect to happen in this case. 

What Optikey does in this scenario is decompose the word, delete the whole word and then recreate everything but the last character. You can see this happening in the following, where I move the mouse before using "BackOne". Does this match up with your experience?

https://user-images.githubusercontent.com/12151404/161027819-754a1017-ba6c-429c-981c-3af90b0906d5.mp4

**Comment by lsaranto** [comment]
> What Optikey does in this scenario is decompose the word, delete the whole word and then recreate everything but the last character. You can see this happening in the following, where I move the mouse before using "BackOne". Does this match up with your experience?

@kmcnaught That looks really similar to what I've experienced. The Finnish language (which is my first language) uses diacritics so I can see it happening. I can't be sure if I've only experienced it with the Finnish language or with the English language too.

**Comment by kmcnaught** [comment]
@lsaranto in the new pre-release 3.2.3 there is an new setting `Limit "BackOne" deletion to a single character` in the Words tab of the Management Console which turns off the unicode decomposition strategy, so that `BackOne` will just function as the backspace key on a physical keyboard. 

https://github.com/OptiKey/OptiKey/releases

Some notes on our discussion in case we consider this topic again in the future...

For latin script with diacritics turning on this setting should give you a reasonable (not unexpected) behaviour. The more complicated behaviour is necessary for supporting logographic languages like Korean, and may also be preferable in languages with keyboards that apply diacritics to existing characters (i.e. using the `a` key followed by `umlaut` key). It depends on language, use-case and whether the application someone is typing into does its own decomposition when you press backspace. It may be that we could improve the unicode-normalisation logic so that we only change the end of the string - i.e. decompose the last character only. This should work fine with latin-ish languages, but might not behave correctly in Korean. If we made this change we would have to verify each of the languages.

**Comment by kmcnaught** [comment]
I'm closing this, with the caveat that we are still considering whether we can improve the scratchpad's contextual relevance along the lines of https://github.com/OptiKey/OptiKey/issues/453, which would also help avoid/minimise this unexpected behaviour
